CARET_trace/include/caret_trace/recordable_data.hpp

```diff
@@ -0,0 +1,335 @@
+  /// @param trace_point Trace point name for this instance.
+  explicit RecordableData(char * trace_point) : RecordableData(std::string(trace_point)) {}
+
+  ~RecordableData() override {}
+
+  /// @brief Assign recording function.
+  /// @param function recording function.
+  void assign(StdFuncT function)
+  {
+    std::lock_guard<std::shared_mutex> lock(mutex_);
+
+    func_ = function;
+  }
+
+  void start() override
+  {
+    std::lock_guard<std::shared_mutex> lock(mutex_);
+
+    reset_();
+
+    it_ = set_.begin();
+    is_iterating_ = true;  // transition to PREPARE state
+    is_end__ = is_end_iterator_();
+  }
+
+  /// @brief Store new data.
+  /// @param data to store.
+  /// @return True, data was stored to pending set.
+  /// @return False, data was stored to set.
+  bool store(Args... args)
+  {
+    std::lock_guard<std::shared_mutex> lock(mutex_);
+
+    if (is_iterating_) {
+      // PREPARE state
+      pending_set_.insert(args...);
+      return true;
+    } else {
+      // OTHER state
+      set_.insert(args...);
+      return false;
+    }
+  }
+
+  void reset() override
+  {
+    std::lock_guard<std::shared_mutex> lock(mutex_);
+
+    reset_();
+  }
+
+  /// @brief Check whether recording function is assigned.
+  /// @return True if recording function is assigned, false otherwise.
+  bool is_assigned() const
+  {
+    std::shared_lock<std::shared_mutex> lock(mutex_);
+
+    return static_cast<bool>(func_);
+  }
+
+  size_t size() const override
+  {
+    std::shared_lock<std::shared_mutex> lock(mutex_);
+
+    return set_.size();
+  }
+
+  size_t pending_size() const override
+  {
+    std::shared_lock<std::shared_mutex> lock(mutex_);
+
+    return pending_set_.size();
+  }
+
+  bool finished() const override
+  {
+    std::shared_lock<std::shared_mutex> lock(mutex_);  // read lock
+
+    return is_end__;
+  }
+
+  bool is_recording() const override
+  {
+    std::shared_lock<std::shared_mutex> lock(mutex_);
+
+    return is_iterating_;
+  }
+
+  void record_next_one() override
+  {
+    std::lock_guard<std::shared_mutex> lock(mutex_);
+
+    record_one_();
+    next_();
+    try_merge_pending_data();
+  }
+
+  /// @brief Det iterating data.
+  /// @return Current iterating data.
+  const KeyT & get() const
+  {
+    assert(!finished());
+
+    std::shared_lock<std::shared_mutex> lock(mutex_);
+    return *it_;
+  }
+
+  const std::string & trace_point() const override
+  {
+    std::shared_lock<std::shared_mutex> lock(mutex_);
+
+    return trace_point_;
+  }
+
+private:
+  using SetT = KeysSet<Args...>;
+  using ConstIteratorT = typename KeysSet<Args...>::ConstIteratorT;
+
+  void merge_pending_keys()
+  {
+    for (auto & key : pending_set_) {
+      set_.insert(key);
+    }
+    pending_set_.clear();
+  }
+
+  /// @brief merge_pending_data when iterator reached to the end.
+  /// @return true : Succeed to merge.
+  /// @return false : No data were merged.
+  bool try_merge_pending_data()
+  {
+    if (is_end__) {
```

Comment:
Please add comments what is end

Reply:
addressed: [7f3f87e](https://github.com/tier4/CARET_trace/pull/68/commits/7f3f87ea264aab99006074c1bc55375264dc117e)